### ISSUE TITLE
Packer's use of the aws-sdk-go failing to authenticate

### ISSUE THREAD

**Issue by martinb3** [hashicorp/packer#4279]
Hello! We've been trying to use packer inside AWS' new "CodeBuild" product, and we're noticing that while the AWS CLI can authenticate successfully to various EC2 APIs, packer cannot. When we run packer, we get:

> Build 'amazon-ebs' errored: Error querying AMI: NoCredentialProviders: no valid providers in chain. Deprecated.

Since CodeBuild uses ECS containers, where we've assigned the appropriate IAM roles, our working theory is that packer and/or the Go SDK is failing to pull credentials from `169.254.170.2/$AWS_CONTAINER_CREDENTIALS_RELATIVE_URI`. When we manually curl that URL, and parse out credentials and put them in `~/.aws/credentials`, packer begins working.

According to [the SDK docs](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html), this should be happening automatically:

> If your container instance is using at least version 1.11.0 of the container agent and a supported version of the AWS CLI or SDKs, then the SDK client will see that the AWS_CONTAINER_CREDENTIALS_RELATIVE_URI variable is available, and it will use the provided credentials to make calls to the AWS APIs.

And [from what I can tell](https://github.com/aws/aws-sdk-go/blob/master/aws/defaults/defaults.go), this should be happening in the Go SDK.

- Packer version: 0.12.0

- Host platform: Ubuntu 14.04 x64, aws/codebuild/ubuntu-base:14.04 from http://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref.html

- Debug log output from `PACKER_LOG=1 packer build template.json`.
https://gist.github.com/martinb3/ed7e11d33915e7fae49199bf7f479da9

- The _simplest example template and scripts_ needed to reproduce the bug.

We're actually using a very, very simple buildspec.yml:
```
version: 0.1

environment_variables:
  plaintext:
    PACKER_NO_COLOR: "1"
    PACKER_LOG: "1"

phases:
  install:
    commands:
      - wget https://releases.hashicorp.com/packer/0.12.0/packer_0.12.0_linux_amd64.zip
      - unzip packer_0.12.0_linux_amd64.zip
  build:
    commands:
      - ./packer build "ami-stuff/packer.json"
```

The packer.json file is also very simple:
```
{
  "builders": [{
    "type": "amazon-ebs",
    "region": "us-east-1",
    "source_ami": "ami-72343365",
    "instance_type": "t2.micro",
    "ssh_username": "ubuntu",
    "ssh_pty" : true,
    "ami_name": "ami-builder-demo {{timestamp}}"
  }],
  "provisioners": [{
    "type": "shell",
    "inline": [
      "sleep 30",
      "sudo apt-get update",
      "sudo apt-get install apache2 -y"
    ]
  }]
}
```

Our workaround has been to setup the credentials file manually:
```
curl -qL 169.254.170.2/$AWS_CONTAINER_CREDENTIALS_RELATIVE_URI > credentials.json
mkdir -p ~/.aws
echo "[default]" > ~/.aws/credentials
echo "aws_access_key_id = $(cat credentials.json | jq '.AccessKeyId' | sed s'/"//g')" >> ~/.aws/credentials
echo "aws_secret_access_key = $(cat credentials.json | jq '.SecretAccessKey' | sed s'/"//g')" >> ~/.aws/credentials
echo "aws_session_token = $(cat credentials.json | jq '.Token' | sed s'/"//g')" >> ~/.aws/credentials
```

The entire time, even while packer is failing, commands like `aws ec2 describe-instances` work, so we know the CLI and Python SDK are correctly using that URL above to pull credentials, even when packer and/or the Go SDK isn't.

**Comment by rickard-von-essen** [comment]
Yes we currently don't support task-roles. I started to look into fixing this but I don't think it will happen before the holidays.

**Comment by rickard-von-essen** [comment]
Related to #4124

**Comment by rickard-von-essen** [comment]
This will be fixed by #4613. Unfortunately at the moment you are not allowed to SSH out of a CodeBuild container, see [this thread](https://forums.aws.amazon.com/message.jspa?messageID=757526). But as soon as that is allowed Packer should work when #4613 have been included.

**Comment by martinb3** [comment]
I think that thread is specific to SSH to accessing EC2 instances on private networks. I had successfully worked around this issue and built AMIs using Packer already, with public IPs on EC2 instances, just by implementing the workaround mentioned above.

I'm happy to hear I might be able to take out the workaround -- thanks for your hard work on this! #4613 looks like a big change 👍

**Comment by 2rs2ts** [comment]
For those curious, this is unrelated to ECS or CodeBuild. We are getting these problems with just using the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables starting with version 0.12.0. (they work fine in 0.11.0.) Our Jenkins builders do not have an `.aws` directory in their homes so we get an error that says:

```
SharedConfigLoadError: failed to load config file, /var/lib/jenkins/.aws/credentials
```

Of course the PR linked fixes our problem.

**Comment by mwhooker** [comment]
fixed in #4996
